// cliui/build/lib/index.js
2: 0,  # 'ร'
        61: 0,  # 'ฤ'
        15: 0,  # 'ล'
        12: 0,  # 'ว'
        42: 0,  # 'ศ'
        46: 0,  # 'ษ'
        18: 3,  # 'ส'
        21: 3,  # 'ห'
        4: 0,  # 'อ'
        63: 0,  # 'ฯ'
        22: 0,  # 'ะ'
        10: 0,  # 'ั'
        1: 0,  # 'า'
        36: 0,  # 'ำ'
        23: 0,  # 'ิ'
        13: 0,  # 'ี'
        40: 0,  # 'ึ'
        27: 0,  # 'ื'
        32: 0,  # 'ุ'
        35: 0,  # 'ู'
        11: 0,  # 'เ'
        28: 0,  # 'แ'
        41: 0,  # 'โ'
        29: 0,  # 'ใ'
        33: 0,  # 'ไ'
        50: 0,  # 'ๆ'
        37: 0,  # '็'
        6: 0,  # '่'
        7: 0,  # '้'
        38: 0,  # '์'
        56: 0,  # '๑'
        59: 0,  # '๒'
        60: 0,  # '๕'
    },
    33: {  # 'ไ'
        5: 1,  # 'ก'
        30: 2,  # 'ข'
        24: 0,  # 'ค'
        8: 0,  # 'ง'
        26: 0,  # 'จ'
        52: 0,  # 'ฉ'
        34: 1,  # 'ช'
        51: 1,  # 'ซ'
        47: 0,  # 'ญ'
        58: 0,  # 'ฎ'
        57: 0,  # 'ฏ'
        49: 0,  # 'ฐ'
        53: 0,  # 'ฑ'
        55: 0,  # 'ฒ'
        43: 0,  # 'ณ'
        20: 3,  # 'ด'
        19: 1,  # 'ต'
        44: 0,  # 'ถ'
        14: 3,  # 'ท'
        48: 0,  # 'ธ'
        3: 0,  # 'น'
        17: 1,  # 'บ'
        25: 3,  # 'ป'
        39: 0,  # 'ผ'
        62: 0,  # 'ฝ'
        31: 0,  # 'พ'
        54: 2,  # 'ฟ'
        45: 0,  # 'ภ'
        9: 3,  # 'ม'
        16: 0,  # 'ย'
        2: 3,  # 'ร'
        61: 0,  # 'ฤ'
        15: 1,  # 'ล'
        12: 3,  # 'ว'
        42: 0,  # 'ศ'
        46: 0,  # 'ษ'
        18: 1,  # 'ส'
        21: 2,  # 'ห'
        4: 0,  # 'อ'
        63: 0,  # 'ฯ'
        22: 0,  # 'ะ'
        10: 0,  # 'ั'
        1: 0,  # 'า'
        36: 0,  # 'ำ'
        23: 0,  # 'ิ'
        13: 0,  # 'ี'
        40: 0,  # 'ึ'
        27: 0,  # 'ื'
        32: 0,  # 'ุ'
        35: 0,  # 'ู'
        11: 0,  # 'เ'
        28: 0,  # 'แ'
        41: 0,  # 'โ'
        29: 0,  # 'ใ'
        33: 0,  # 'ไ'
        50: 0,  # 'ๆ'
        37: 0,  # '็'
        6: 0,  # '่'
        7: 0,  # '้'
        38: 0,  # '์'
        56: 0,  # '๑'
        59: 0,  # '๒'
        60: 0,  # '๕'
    },
    50: {  # 'ๆ'
        5: 0,  # 'ก'
        30: 0,  # 'ข'
        24: 0,  # 'ค'
        8: 0,  # 'ง'
        26: 0,  # 'จ'
        52: 0,  # 'ฉ'
        34: 0,  # 'ช'
        51: 0,  # 'ซ'
        47: 0,  # 'ญ'
        58: 0,  # 'ฎ'
        57: 0,  # 'ฏ'
        49: 0,  # 'ฐ'
        53: 0,  # 'ฑ'
        55: 0,  # 'ฒ'
        43: 0,  # 'ณ'
        20: 0,  # 'ด'
        19: 0,  # 'ต'
        44: 0,  # 'ถ'
        14: 0,  # 'ท'
        48: 0,  # 'ธ'
        3: 0,  # 'น'
        17: 0,  # 'บ'
        25: 0,  # 'ป'
        39: 0,  # 'ผ'
        62: 0,  # 'ฝ'
        31: 0,  # 'พ'
        54: 0,  # 'ฟ'
        45: 0,  # 'ภ'
        9: 0,  # 'ม'
        16: 0,  # 'ย'
        2: 0,  # 'ร'
        61: 0,  # 'ฤ'
        15: 0,  # 'ล'
        12: 0,  # 'ว'
        42: 0,  # 'ศ'
        46: 0,  # 'ษ'
        18: 0,  # 'ส'
        21: 0,  # 'ห'
        4: 0,  # 'อ'
        63: 0,  # 'ฯ'
        22: 0,  # 'ะ'
        10: 0,  # 'ั'
        1: 0,  # 'า'
        36: 0,  # 'ำ'
        23: 0,  # 'ิ'
        13: 0,  # 'ี'
        40: 0,  # 'ึ'
        27: 0,  # 'ื'
        32: 0,  # 'ุ'
        35: 0,  # 'ู'
        11: 0,  # 'เ'
        28: 0,  # 'แ'
        41: 0,  # 'โ'
        29: 0,  # 'ใ'
        33: 0,  # 'ไ'
        50: 0,  # 'ๆ'
        37: 0,  # '็'
        6: 0,  # '่'
        7: 0,  # '้'
        38: 0,  # '์'
        56: 0,  # '๑'
        59: 0,  # '๒'
        60: 0,  # '๕'
    },
    37: {  # '็'
        5: 2,  # 'ก'
        30: 1,  # 'ข'
        24: 2,  # 'ค'
        8: 2,  # 'ง'
        26: 3,  # 'จ'
        52: 0,  # 'ฉ'
        34: 0,  # 'ช'
        51: 0,  # 'ซ'
        47: 1,  # 'ญ'
        58: 0,  # 'r(row[c], ts, '| ');
                str += ts;
                str += addBorder(row[c], ts, ' |');
                if (padding[right]) {
                    str += ' '.repeat(padding[right]);
                }
                // if prior row is span, try to render the
                // current row on the prior line.
                if (r === 0 && lines.length > 0) {
                    str = this.renderInline(str, lines[lines.length - 1]);
                }
            });
            // remove trailing whitespace.
            lines.push({
                text: str.replace(/ +$/, ''),
                span: row.span
            });
        });
        return lines;
    }
    // if the full 'source' can render in
    // the target line, do so.
    renderInline(source, previousLine) {
        const match = source.match(/^ */);
        const leadingWhitespace = match ? match[0].length : 0;
        const target = previousLine.text;
        const targetTextWidth = mixin.stringWidth(target.trimRight());
        if (!previousLine.span) {
            return source;
        }
        // if we're not applying wrapping logic,
        // just always append to the span.
        if (!this.wrap) {
            previousLine.hidden = true;
            return target + source;
        }
        if (leadingWhitespace < targetTextWidth) {
            return source;
        }
        previousLine.hidden = true;
        return target.trimRight() + ' '.repeat(leadingWhitespace - targetTextWidth) + source.trimLeft();
    }
    rasterize(row) {
        const rrows = [];
        const widths = this.columnWidths(row);
        let wrapped;
        // word wrap all columns, and create
        // a data-structure that is easy to rasterize.
        row.forEach((col, c) => {
            // leave room for left and right padding.
            col.width = widths[c];
            if (this.wrap) {
                wrapped = mixin.wrap(col.text, this.negatePadding(col), { hard: true }).split('\n');
            }
            else {
                wrapped = col.text.split('\n');
            }
            if (col.border) {
                wrapped.unshift('.' + '-'.repeat(this.negatePadding(col) + 2) + '.');
                wrapped.push("'" + '-'.repeat(this.negatePadding(col) + 2) + "'");
            }
            // add top and bottom padding.
            if (col.padding) {
                wrapped.unshift(...new Array(col.padding[top] || 0).fill(''));
                wrapped.push(...new Array(col.padding[bottom] || 0).fill(''));
            }
            wrapped.forEach((str, r) => {
                if (!rrows[r]) {
                    rrows.push([]);
                }
                const rrow = rrows[r];
                for (let i = 0; i < c; i++) {
                    if (rrow[i] === undefined) {
                        rrow.push('');
                    }
                }
                rrow.push(str);
            });
        });
        return rrows;
    }
    negatePadding(col) {
        let wrapWidth = col.width || 0;
        if (col.padding) {
            wrapWidth -= (col.padding[left] || 0) + (col.padding[right] || 0);
        }
        if (col.border) {
            wrapWidth -= 4;
        }
        return wrapWidth;
    }
    columnWidths(row) {
        if (!this.wrap) {
            return row.map(col => {
                return col.width || mixin.stringWidth(col.text);
            });
        }
        let unset = row.length;
        let remainingWidth = this.width;
        // column widths can be set in config.
        const widths = row.map(col => {
            if (col.width) {
                unset--;
                remainingWidth -= col.width;
                return col.width;
            }
            return undefined;
        });
        // any unset widths should be calculated.
        const unsetWidth = unset ? Math.floor(remainingWidth / unset) : 0;
        return widths.map((w, i) => {
            if (w === undefined) {
                return Math.max(unsetWidth, _minWidth(row[i]));
            }
            return w;
        });
    }
}
function addBorder(col, ts, style) {
    if (col.border) {
        if (/[.']-+[.']/.test(ts)) {
            return '';
        }
        if (ts.trim().length !== 0) {
            return style;
        }
        return '  ';
    }
    return '';
}
// calculates the minimum width of
// a column, based on padding preferences.
function _minWidth(col) {
    const padding = col.padding || [];
    const minWidth = 1 + (padding[left] || 0) + (padding[right] || 0);
    if (col.border) {
        return minWidth + 4;
    }
    return minWidth;
}
function getWindowWidth() {
    /* istanbul ignore next: depends on terminal */
    if (typeof process === 'object' && process.stdout && process.stdout.columns) {
        return process.stdout.columns;
    }
    return 80;
}
function alignRight(str, width) {
    str = str.trim();
    const strWidth = mixin.stringWidth(str);
    if (strWidth < width) {
        return ' '.repeat(width - strWidth) + str;
    }
    return str;
}
function alignCenter(str, width) {
    str = str.trim();
    const strWidth = mixin.stringWidth(str);
    /* istanbul ignore next */
    if (strWidth >= width) {
        return str;
    }
    return ' '.repeat((width - strWidth) >> 1) + str;
}
let mixin;
export function cliui(opts, _mixin) {
    mixin = _mixin;
    return new UI({
        width: (opts === null || opts === void 0 ? void 0 : opts.width) || getWindowWidth(),
        wrap: opts === null || opts === void 0 ? void 0 : opts.wrap
    });
}
