// electron-to-chromium/full-chromium-versions.js
module.exports = {
	"39.0.2171.65": [
		"0.20.0",
		"0.20.1",
		"0.20.2",
		"0.20.3",
		"0.20.4",
		"0.20.5",
		"0.20.6",
		"0.20.7",
		"0.20.8"
	],
	"40.0.2214.91": [
		"0.21.0",
		"0.21.1",
		"0.21.2"
	],
	"41.0.2272.76": [
		"0.21.3",
		"0.22.1",
		"0.22.2",
		"0.22.3",
		"0.23.0",
		"0.24.0"
	],
	"42.0.2311.107": [
		"0.25.0",
		"0.25.1",
		"0.25.2",
		"0.25.3",
		"0.26.0",
		"0.26.1",
		"0.27.0",
		"0.27.1"
	],
	"43.0.2357.65": [
		"0.27.2",
		"0.27.3",
		"0.28.0",
		"0.28.1",
		"0.28.2",
		"0.28.3",
		"0.29.1",
		"0.29.2"
	],
	"44.0.2403.125": [
		"0.30.4",
		"0.31.0"
	],
	"45.0.2454.85": [
		"0.31.2",
		"0.32.2",
		"0.32.3",
		"0.33.0",
		"0.33.1",
		"0.33.2",
		"0.33.3",
		"0.33.4",
		"0.33.6",
		"0.33.7",
		"0.33.8",
		"0.33.9",
		"0.34.0",
		"0.34.1",
		"0.34.2",
		"0.34.3",
		"0.34.4",
		"0.35.1",
		"0.35.2",
		"0.35.3",
		"0.35.4",
		"0.35.5"
	],
	"47.0.2526.73": [
		"0.36.0",
		"0.36.2",
		"0.36.3",
		"0.36.4"
	],
	"47.0.2526.110": [
		"0.36.5",
		"0.36.6",
		"0.36.7",
		"0.36.8",
		"0.36.9",
		"0.36.10",
		"0.36.11",
		"0.36.12"
	],
	"49.0.2623.75": [
		"0.37.0",
		"0.37.1",
		"0.37.3",
		"0.37.4",
		"0.37.5",
		"0.37.6",
		"0.37.7",
		"0.37.8",
		"1.0.0",
		"1.0.1",
		"1.0.2"
	],
	"50.0.2661.102": [
		"1.1.0",
		"1.1.1",
		"1.1.2",
		"1.1.3"
	],
	"51.0.2704.63": [
		"1.2.0",
		"1.2.1"
	],
	"51.0.2704.84": [
		"1.2.2",
		"1.2.3"
	],
	"51.0.2704.103": [
		"1.2.4",
		"1.2.5"
	],
	"51.0.2704.106": [
		"1.2.6",
		"1.2.7",
		"1.2.8"
	],
	"52.0.2743.82": [
		"1.3.0",
		"1.3.1",
		"1.3.2",
		"1.3.3",
		"1.3.4",
		"1.3.5",
		"1.3.6",
		"1.3.7",
		"1.3.9",
		"1.3.10",
		"1.3.13",
		"1.3.14",
		"1.3.15"
	],
	"53.0.2785.113": [
		"1.4.0",
		"1.4.1",
		"1.4.2",
		"1.4.3",
		"1.4.4",
		"1.4.5"
	],
	"53.0.2785.143": [
		"1.4.6",
		"1.4.7",
		"1.4.8",
		"1.4.10",
		"1.4.11",
		"1.4.13",
		"1.4.14",
		"1.4.15",
		"1.4.16"
	],
	"54.0.2840.51": [
		"1.4.12"
	],
	"54.0.2840.101": [
		"1.5.0",
		"1.5.1"
	],
	"56.0.2924.87": [
		"1.6.0",
		"1.6.1",
		"1.6.2",
		"1.6.3",
		"1.6.4",
		"1.6.5",
		"1.6.6",
		"1.6.7",
		"1.6.8",
		"1.6.9",
		"1.6.10",
		"1.6.11",
		"1.6.12",
		"1.6.13",
		"1.6.14",
		"1.6.15",
		"1.6.16",
		"1.6.17",
		"1.6.18"
	],
	"58.0.3029.110": [
		"1.7.0",
		"1.7.1",
		"1.7.2",
		"1.7.3",
		"1.7.4",
		"1.7.5",
		"1.7.6",
		"1.7.7",
		"1.7.8",
		"1.7.9",
		"1.7.10",
		"1.7.11",
		"1.7.12",
		"1.7.13",
		"1.7.14",
		"1.7.15",
		"1.7.16"
	],
	"59.0.3071.115": [
		"1.8.0",
		"1.8.1",
		"1.8.2-beta.1",
		"1.8.2-beta.2",
		"1.8.2-beta.3",
		"1.8.2-beta.4",
		"1.8.2-beta.5",
		"1.8.2",
		"1.8.3",
		"1.8.4",
		"1.8.5",
		"1.8.6",
		"1.8.7",
		"1.8.8"
	],
	"61.0.3163.100": [
		"2.0.0-beta.1",
		"2.0.0-beta.2",
		"2.0.0-beta.3",
		"2.0.0-beta.4",
		"2.0.0-beta.5",
		"2.0.0-beta.6",
		"2.0.0-beta.7",
		"2.0.0-beta.8",
		"2.0.0",
		"2.0.1",
		"2.0.2",
		"2.0.3",
		"2.0.4",
		"2.0.5",
		"2.0.6",
		"2.0.7",
		"2.0.8",
		"2.0.9",
		"2.0.10",
		"2.0.11",
		"2.0.12",
		"2.0.13",
		"2.0.14",
		"2.0.15",
		"2.0.16",
		"2.0.17",
		"2.0.18",
		"2.1.0-unsupported.20180809"
	],
	"66.0.3359.181": [
		"3.0.0-beta.1",
		"3.0.0-beta.2",
		"3.0.0-beta.3",
		"3.0.0-beta.4",
		"3.0.0-beta.5",
		"3.0.0-beta.6",
		"3.0.0-beta.7",
		"3.0.0-beta.8",
		"3.0.0-beta.9",
		"3.0.0-beta.10",
		"3.0.0-beta.11",
		"3.0.0-beta.12",
		"3.0.0-beta.13",
		"3.0.0",
		"3.0.1",
		"3.0.2",
		"3.0.3",
		"3.0.4",
		"3.0.5",
		"3.0.6",
		"3.0.7",
		"3.0.8",
		"3.0.9",
		"3.0.10",
		"3.0.11",
		"3.0.12",
		"3.0.13",
		"3.0.14",
		"3.0.15",
		"3.0.16",
		"3.1.0-beta.1",
		"3.1.0-beta.2",
		"3.1.0-beta.3",
		"3.1.0-beta.4",
		"3.1.0-beta.5",
		"3.1.0",
		"3.1.1",
		"3.1.2",
		"3.1.3",
		"3.1.4",
		"3.1.5",
		"3.1.6",
		"3.1.7",
		"3.1.8",
		"3.1.9",
		"3.1.10",
		"3.1.11",
		"3.1.12",
		"3.1.13"
	],
	"69.0.3497.106": [
		"4.0.0-beta.1",
		"4.0.0-beta.2",
		"4.0.0-beta.3",
		"4.0.0-beta.4",
		"4.0.0-beta.5",
		"4.0.0-beta.6",
		"4.0.0-beta.7",
		"4.0.0-beta.8",
		"4.0.0-beta.9",
		"4.0.0-beta.10",
		"4.0.0-beta.11",
		"4.0.0",
		"4.0.1",
		"4.0.2",
		"4.0.3",
		"4.0.4",
		"4.0.5",
		"4.0.6"
	],
	"69.0.3497.128": [
		"4.0.7",
		"4.0.8",
		"4.1.0",
		"4.1.1",
		"4.1.2",
		"4.1.3",
		"4.1.4",
		"4.1.5",
		"4.2.0",
		"4.2.1",
		"4.2.2",
		"4.2.3",
		"4.2.4",
		"4.2.5",
		"4.2.6",
		"4.2.7",
		"4.2.8",
		"4.2.9",
		"4.2.10",
		"4.2.11",
		"4.2.12"
	],
	"72.0.3626.52": [
		"5.0.0-beta.1",
		"5.0.0-beta.2"
	],
	"73.0.3683.27": [
		"5.0.0-beta.3"
	],
	"73.0.3683.54": [
		"5.0.0-beta.4"
	],
	"73.0.3683.61": [
		"5.0.0-beta.5"
	],
	"73.0.3683.84": [
		"5.0.0-beta.6"
	],
	"73.0.3683.94": [
		"5.0.0-beta.7"
	],
	"73.0.3683.104": [
		"5.0.0-beta.8"
	],
	"73.0.3683.117": [
		"5.0.0-beta.9"
	],
	"73.0.3683.119": [
		"5.0.0"
	],
	"73.0.3683.121": [
		"5.0.1",
		"5.0.2",
		"5.0.3",
		"5.0.4",
		"5.0.5",
		"5.0.6",
		"5.0.7",
		"5.0.8",
		"5.0.9",
		"5.0.10",
		"5.0.11",
		"5.0.12",
		"5.0.13"
	],
	"76.0.3774.1": [
		"6.0.0-beta.1"
	],
	"76.0.3783.1": [
		"6.0.0-beta.2",
		"6.0.0-beta.3",
		"6.0.0-beta.4"
	],
	"76.0.3805.4": [
		"6.0.0-beta.5"
	],
	"76.0.3809.3": [
		"6.0.0-beta.6"
	],
	"76.0.3809.22": [
		"6.0.0-beta.7"
	],
	"76.0.3809.26": [
		"6.0.0-beta.8",
		"6.0.0-beta.9"
	],
	"76.0.3809.37": [
		"6.0.0-beta.10"
	],
	"76.0.3809.42": [
		"6.0.0-beta.11"
	],
	"76.0.3809.54": [
		"6.0.0-beta.12"
	],
	"76.0.3809.60": [
		"6.0.0-beta.13"
	],
	"76.0.3809.68": [
		"6.0.0-beta.14"
	],
	"76.0.3809.74": [
		"6.0.0-beta.15"
	],
	"76.0.3809.88": [
		"6.0.0"
	],
	"76.0.3809.102": [
		"6.0.1"
	],
	"76.0.3809.110": [
		"6.0.2"
	],
	"76.0.3809.126": [
		"6.0.3"
	],
	"76.0.3809.131": [
		"6.0.4"
	],
	"76.0.3809.136": [
		"6.0.5"
	],
	"76.0.3809.138": [
		"6.0.6"
	],
	"76.0.3809.139": [
		"6.0.7"
	],
	"76.0.3809.146": [
		"6.0.8",
		"6.0.9",
		"6.0.10",
		"6.0.11",
		"6.0.12",
		"6.1.0",
		"6.1.1",
		"6.1.2",
		"6.1.3",
		"6.1.4",
		"6.1.5",
		"6.1.6",
		"6.1.7",
		"6.1.8",
		"6.1.9",
		"6.1.10",
		"6.1.11",
		"6.1.12"
	],
	"78.0.3866.0": [
		"7.0.0-beta.1",
		"7.0.0-beta.2",
		"7.0.0-beta.3"
	],
	"78.0.3896.6": [
		"7.0.0-beta.4"
	],
	"78.0.3905.1": [
		"7.0.0-beta.5",
		"7.0.0-beta.6",
		"7.0.0-beta.7",
		"7.0.0"
	],
	"78.0.3904.92": [
		"7.0.1"
	],
	"78.0.3904.94": [
		"7.1.0"
	],
	"78.0.3904.99": [
		"7.1.1"
	],
	"78.0.3904.113": [
		"7.1.2"
	],
	"78.0.3904.126": [
		"7.1.3"
	],
	"78.0.3904.130": [
		"7.1.4",
		"7.1.5",
		"7.1.6",
		"7.1.7",
		"7.1.8",
		"7.1.9",
		"7.1.10",
		"7.1.11",
		"7.1.12",
		"7.1.13",
		"7.1.14",
		"7.2.0",
		"7.2.1",
		"7.2.2",
		"7.2.3",
		"7.2.4",
		"7.3.0",
		"7.3.1",
		"7.3.2",
		"7.3.3"
	],
	"79.0.3931.0": [
		"8.0.0-beta.1",
		"8.0.0-beta.2"
	],
	"80.0.3955.0": [
		"8.0.0-beta.3",
		"8.0.0-beta.4"
	],
	"80.0.3987.14": [
		"8.0.0-beta.5"
	],
	"80.0.3987.51": [
		"8.0.0-beta.6"
	],
	"80.0.3987.59": [
		"8.0.0-beta.7"
	],
	"80.0.3987.75": [
		"8.0.0-beta.8",
		"8.0.0-beta.9"
	],
	"80.0.3987.86": [
		"8.0.0",
		"8.0.1",
		"8.0.2"
	],
	"80.0.3987.134": [
		"8.0.3"
	],
	"80.0.3987.137": [
		"8.1.0"
	],
	"80.0.3987.141": [
		"8.1.1"
	],
	"80.0.3987.158": [
		"8.2.0"
	],
	"80.0.3987.163": [
		"8.2.1",
		"8.2.2",
		"8.2.3",
		"8.5.3",
		"8.5.4",
		"8.5.5"
	],
	"80.0.3987.165": [
		"8.2.4",
		"8.2.5",
		"8.3.0",
		"8.3.1",
		"8.3.2",
		"8.3.3",
		"8.3.4",
		"8.4.0",
		"8.4.1",
		"8.5.0",
		"8.5.1",
		"8.5.2"
	],
	"82.0.4048.0": [
		"9.0.0-beta.1",
		"9.0.0-beta.2",
		"9.0.0-beta.3",
		"9.0.0-beta.4",
		"9.0.0-beta.5"
	],
	"82.0.4058.2": [
		"9.0.0-beta.6",
		"9.0.0-beta.7",
		"9.0.0-beta.9"
	],
	"82.0.4085.10": [
		"9.0.0-beta.10"
	],
	"82.0.4085.14": [
		"9.0.0-beta.11",
		"9.0.0-beta.12",
		"9.0.0-beta.13"
	],
	"82.0.4085.27": [
		"9.0.0-beta.14"
	],
	"83.0.4102.3": [
		"9.0.0-beta.15",
		"9.0.0-beta.16"
	],
	"83.0.4103.14": [
		"9.0.0-beta.17"
	],
	"83.0.4103.16": [
		"9.0.0-beta.18"
	],
	"83.0.4103.24": [
		"9.0.0-beta.19"
	],
	"83.0.4103.26": [
		"9.0.0-beta.20",
		"9.0.0-beta.21"
	],
	"83.0.4103.34": [
		"9.0.0-beta.22"
	],
	"83.0.4103.44": [
		"9.0.0-beta.23"
	],
	"83.0.4103.45": [
		"9.0.0-beta.24"
	],
	"83.0.4103.64": [
		"9.0.0"
	],
	"83.0.4103.94": [
		"9.0.1",
		"9.0.2"
	],
	"83.0.4103.100": [
		"9.0.3"
	],
	"83.0.4103.104": [
		"9.0.4"
	],
	"83.0.4103.119": [
		"9.0.5"
	],
	"83.0.4103.122": [
		"9.1.0",
		"9.1.1",
		"9.1.2",
		"9.2.0",
		"9.2.1",
		"9.3.0",
		"9.3.1",
		"9.3.2",
		"9.3.3",
		"9.3.4",
		"9.3.5",
		"9.4.0",
		"9.4.1",
		"9.4.2",
		"9.4.3",
		"9.4.4"
	],
	"84.0.4129.0": [
		"10.0.0-beta.1",
		"10.0.0-beta.2"
	],
	"85.0.4161.2": [
		"10.0.0-beta.3",
		"10.0.0-beta.4"
	],
	"85.0.4181.1": [
		"10.0.0-beta.8",
		"10.0.0-beta.9"
	],
	"85.0.4183.19": [
		"10.0.0-beta.10"
	],
	"85.0.4183.20": [
		"10.0.0-beta.11"
	],
	"85.0.4183.26": [
		"10.0.0-beta.12"
	],
	"85.0.4183.39": [
		"10.0.0-beta.13",
		"10.0.0-beta.14",
		"10.0.0-beta.15",
		"10.0.0-beta.17",
		"10.0.0-beta.19",
		"10.0.0-beta.20",
		"10.0.0-beta.21"
	],
	"85.0.4183.70": [
		"10.0.0-beta.23"
	],
	"85.0.4183.78": [
		"10.0.0-beta.24"
	],
	"85.0.4183.80": [
		"10.0.0-beta.25"
	],
	"85.0.4183.84": [
		"10.0.0"
	],
	"85.0.4183.86": [
		"10.0.1"
	],
	"85.0.4183.87": [
		"10.1.0"
	],
	"85.0.4183.93": [
		"10.1.1"
	],
	"85.0.4183.98": [
		"10.1.2"
	],
	"85.0.4183.121": [
		"10.1.3",
		"10.1.4",
		"10.1.5",
		"10.1.6",
		"10.1.7",
		"10.2.0",
		"10.3.0",
		"10.3.1",
		"10.3.2",
		"10.4.0",
		"10.4.1",
		"10.4.2",
		"10.4.3",
		"10.4.4",
		"10.4.5",
		"10.4.6",
		"10.4.7"
	],
	"86.0.4234.0": [
		"11.0.0-beta.1",
		"11.0.0-beta.3",
		"11.0.0-beta.4",
		"11.0.0-beta.5",
		"11.0.0-beta.6",
		"11.0.0-beta.7"
	],
	"87.0.4251.1": [
		"11.0.0-beta.8",
		"11.0.0-beta.9",
		"11.0.0-beta.11"
	],
	"87.0.4280.11": [
		"11.0.0-beta.12",
		"11.0.0-beta.13"
	],
	"87.0.4280.27": [
		"11.0.0-beta.16",
		"11.0.0-beta.17",
		"11.0.0-beta.18",
		"11.0.0-beta.19"
	],
	"87.0.4280.40": [
		"11.0.0-beta.20"
	],
	"87.0.4280.47": [
		"11.0.0-beta.22",
		"11.0.0-beta.23"
	],
	"87.0.4280.60": [
		"11.0.0",
		"11.0.1"
	],
	"87.0.4280.67": [
		"11.0.2",
		"11.0.3",
		"11.0.4"
	],
	"87.0.4280.88": [
		"11.0.5",
		"11.1.0",
		"11.1.1"
	],
	"87.0.4280.141": [
		"11.2.0",
		"11.2.1",
		"11.2.2",
		"11.2.3",
		"11.3.0",
		"11.4.0",
		"11.4.1",
		"11.4.2",
		"11.4.3",
		"11.4.4",
		"11.4.5",
		"11.4.6",
		"11.4.7",
		"11.4.8",
		"11.4.9",
		"11.4.10",
		"11.4.11",
		"11.4.12",
		"11.5.0"
	],
	"89.0.4328.0": [
		"12.0.0-beta.1",
		"12.0.0-beta.3",
		"12.0.0-beta.4",
		"12.0.0-beta.5",
		"12.0.0-beta.6",
		"12.0.0-beta.7",
		"12.0.0-beta.8",
		"12.0.0-beta.9",
		"12.0.0-beta.10",
		"12.0.0-beta.11",
		"12.0.0-beta.12",
		"12.0.0-beta.14"
	],
	"89.0.4348.1": [
		"12.0.0-beta.16",
		"12.0.0-beta.18",
		"12.0.0-beta.19",
		"12.0.0-beta.20"
	],
	"89.0.4388.2": [
		"12.0.0-beta.21",
		"12.0.0-beta.22",
		"12.0.0-beta.23",
		"12.0.0-beta.24",
		"12.0.0-beta.25",
		"12.0.0-beta.26"
	],
	"89.0.4389.23": [
		"12.0.0-beta.27",
		"12.0.0-beta.28",
		"12.0.0-beta.29"
	],
	"89.0.4389.58": [
		"12.0.0-beta.30",
		"12.0.0-beta.31"
	],
	"89.0.4389.69": [
		"12.0.0"
	],
	"89.0.4389.82": [
		"12.0.1"
	],
	"89.0.4389.90": [
		"12.0.2"
	],
	"89.0.4389.114": [
		"12.0.3",
		"12.0.4"
	],
	"89.0.4389.128": [
		"12.0.5",
		"12.0.6",
		"12.0.7",
		"12.0.8",
		"12.0.9",
		"12.0.10",
		"12.0.11",
		"12.0.12",
		"12.0.13",
		"12.0.14",
		"12.0.15",
		"12.0.16",
		"12.0.17",
		"12.0.18",
		"12.1.0",
		"12.1.1",
		"12.1.2",
		"12.2.0",
		"12.2.1",
		"12.2.2",
		"12.2.3"
	],
	"90.0.4402.0": [
		"13.0.0-beta.2",
		"13.0.0-beta.3"
	],
	"90.0.4415.0": [
		"13.0.0-beta.4",
		"13.0.0-beta.5",
		"13.0.0-beta.6",
		"13.0.0-beta.7",
		"13.0.0-beta.8",
		"13.0.0-beta.9",
		"13.0.0-beta.10",
		"13.0.0-beta.11",
		"13.0.0-beta.12",
		"13.0.0-beta.13"
	],
	"91.0.4448.0": [
		"13.0.0-beta.14",
		"13.0.0-beta.16",
		"13.0.0-beta.17",
		"13.0.0-beta.18",
		"13.0.0-beta.20"
	],
	"91.0.4472.33": [
		"13.0.0-beta.21",
		"13.0.0-beta.22",
		"13.0.0-beta.23"
	],
	"91.0.4472.38": [
		"13.0.0-beta.24",
		"13.0.0-beta.25",
		"13.0.0-beta.26",
		"13.0.0-beta.27",
		"13.0.0-beta.28"
	],
	"91.0.4472.69": [
		"13.0.0",
		"13.0.1"
	],
	"91.0.4472.77": [
		"13.1.0",
		"13.1.1",
		"13.1.2"
	],
	"91.0.4472.106": [
		"13.1.3",
		"13.1.4"
	],
	"91.0.4472.124": [
		"13.1.5",
		"13.1.6",
		"13.1.7"
	],
	"91.0.4472.164": [
		"13.1.8",
		"13.1.9",
		"13.2.0",
		"13.2.1",
		"13.2.2",
		"13.2.3",
		"13.3.0",
		"13.4.0",
		"13.5.0",
		"13.5.1",
		"13.5.2",
		"13.6.0",
		"13.6.1",
		"13.6.2",
		"13.6.3",
		"13.6.6",
		"13.6.7",
		"13.nore_requires_python,
        )
        build_tracker = self.enter_context(get_build_tracker())

        directory = TempDirectory(
            delete=not options.no_clean,
            kind="install",
            globally_managed=True,
        )

        try:
            reqs = self.get_requirements(args, options, finder, session)
            check_legacy_setup_py_options(options, reqs)

            wheel_cache = WheelCache(options.cache_dir)

            # Only when installing is it permitted to use PEP 660.
            # In other circumstances (pip wheel, pip download) we generate
            # regular (i.e. non editable) metadata and wheels.
            for req in reqs:
                req.permit_editable_wheels = True

            preparer = self.make_requirement_preparer(
                temp_build_dir=directory,
                options=options,
                build_tracker=build_tracker,
                session=session,
                finder=finder,
                use_user_site=options.use_user_site,
                verbosity=self.verbosity,
            )
            resolver = self.make_resolver(
                preparer=preparer,
                finder=finder,
                options=options,
                wheel_cache=wheel_cache,
                use_user_site=options.use_user_site,
                ignore_installed=options.ignore_installed,
                ignore_requires_python=options.ignore_requires_python,
                force_reinstall=options.force_reinstall,
                upgrade_strategy=upgrade_strategy,
                use_pep517=options.use_pep517,
            )

            self.trace_basic_info(finder)

            requirement_set = resolver.resolve(
                reqs, check_supported_wheels=not options.target_dir
            )

            if options.json_report_file:
                report = InstallationReport(requirement_set.requirements_to_install)
                if options.json_report_file == "-":
                    print_json(data=report.to_dict())
                else:
                    with open(options.json_report_file, "w", encoding="utf-8") as f:
                        json.dump(report.to_dict(), f, indent=2, ensure_ascii=False)

            if options.dry_run:
                # In non dry-run mode, the legacy versions and specifiers check
                # will be done as part of conflict detection.
                requirement_set.warn_legacy_versions_and_specifiers()
                would_install_items = sorted(
                    (r.metadata["name"], r.metadata["version"])
                    for r in requirement_set.requirements_to_install
                )
                if would_install_items:
                    write_output(
                        "Would install %s",
                        " ".join("-".join(item) for item in would_install_items),
                    )
                return SUCCESS

            try:
                pip_req = requirement_set.get_requirement("pip")
            except KeyError:
                modifying_pip = False
            else:
                # If we're not replacing an already installed pip,
                # we're not modifying it.
                modifying_pip = pip_req.satisfied_by is None
            protect_pip_from_modification_on_windows(modifying_pip=modifying_pip)

            reqs_to_build = [
                r
                for r in requirement_set.requirements.values()
                if should_build_for_install_command(r)
            ]

            _, build_failures = build(
                reqs_to_build,
                wheel_cache=wheel_cache,
                verify=True,
                build_options=[],
                global_options=global_options,
            )

            if build_failures:
                raise InstallationError(
                    "Could not build wheels for {}, which is required to "
                    "install pyproject.toml-based projects".format(
                        ", ".join(r.name for r in build_failures)  # type: ignore
                    )
                )

            to_install = resolver.get_installation_order(requirement_set)

            # Check for conflicts in the package set we're installing.
            conflicts: Optional[ConflictDetails] = None
            should_warn_about_conflicts = (
                not options.ignore_dependencies and options.warn_about_conflicts
            )
            if should_warn_about_conflicts:
                conflicts = self._determine_conflicts(to_install)

            # Don't warn about script install locations if
            # --target or --prefix has been specified
            warn_script_location = options.warn_script_location
            if options.target_dir or options.prefix_path:
                warn_script_location = False

            installed = install_given_reqs(
                to_install,
                global_options,
                root=options.root_path,
                home=target_temp_dir_path,
                prefix=options.prefix_path,
                warn_script_location=warn_script_location,
                use_user_site=options.use_user_site,
                pycompile=options.compile,
            )

            lib_locations = get_lib_location_guesses(
                user=options.use_user_site,
                home=target_temp_dir_path,
                root=options.root_path,
                prefix=options.prefix_path,
                isolated=options.isolated_mode,
            )
            env = get_environment(lib_locations)

            installed.sort(key=operator.attrgetter("name"))
            items = []
            for result in installed:
                item = result.name
                try:
                    installed_dist = env.get_distribution(item)
                    if installed_dist is not None:
                        item = f"{item}-{installed_dist.version}"
                except Exception:
                    pass
                items.append(item)

            if conflicts is not None:
                self._warn_about_conflicts(
                    conflicts,
                    resolver_variant=self.determine_resolver_variant(options),
                )

            installed_desc = " ".join(items)
            if installed_desc:
                write_output(
                    "Successfully installed %s",
                    installed_desc,
                )
        except OSError as error:
            show_traceback = self.verbosity >= 1

            message = create_os_error_message(
                error,
                show_traceback,
                options.use_user_site,
            )
            logger.error(message, exc_info=show_traceback)

            return ERROR

        if options.target_dir:
            assert target_temp_dir
            self._handle_target_dir(
                options.target_dir, target_temp_dir, options.upgrade
            )
        if options.root_user_action == "warn":
            warn_if_run_as_root()
        return SUCCESS

    def _handle_target_dir(
        self, target_dir: str, target_temp_dir: TempDirectory, upgrade: bool
    ) -> None:
        ensure_dir(target_dir)

        # Checking both purelib and platlib directories for installed
        # packages to be moved to target directory
        lib_dir_list = []

        # Checking both purelib and platlib directories for installed
        # packages to be moved to target directory
        scheme = get_scheme("", home=target_temp_dir.path)
        purelib_dir = scheme.purelib
        platlib_dir = scheme.platlib
        data_dir = scheme.data

        if os.path.exists(purelib_dir):
            lib_dir_list.append(purelib_dir)
        if os.path.exists(platlib_dir) and platlib_dir != purelib_dir:
            lib_dir_list.append(platlib_dir)
        if os.path.exists(data_dir):
            lib_dir_list.append(data_dir)

        for lib_dir in lib_dir_list:
            for item in os.listdir(lib_dir):
                if lib_dir == data_dir:
                    ddir = os.path.join(data_dir, item)
                    if any(s.startswith(ddir) for s in lib_dir_list[:-1]):
                        continue
                target_item_dir = os.path.join(target_dir, item)
                if os.path.exists(target_item_dir):
                    if not upgrade:
                        logger.warning(
                            "Target directory %s already exists. Specify "
                            "--upgrade to force replacement.",
                            target_item_dir,
                        )
                        continue
                    if os.path.islink(target_item_dir):
                        logger.warning(
                            "Target directory %s already exists and is "
                            "a link. pip will not automatically replace "
                            "links, please remove if replacement is "
                            "desired.",
                            target_item_dir,
                        )
                        continue
                    if os.path.isdir(target_item_dir):
                        shutil.rmtree(target_item_dir)
                    else:
                        os.remove(target_item_dir)

                shutil.move(os.path.join(lib_dir, item), target_item_dir)

    def _determine_conflicts(
        self, to_install: List[InstallRequirement]
    ) -> Optional[ConflictDetails]:
        try:
            return check_install_conflicts(to_install)
        except Exception:
            logger.exception(
                "Error while checking for conflicts. Please file an issue on "
                "pip's issue tracker: https://github.com/pypa/pip/issues/new"
            )
            return None

    def _warn_about_conflicts(
        self, conflict_details: ConflictDetails, resolver_variant: str
    ) -> None:
        package_set, (missing, conflicting) = conflict_details
        if not missing and not conflicting:
            return

        parts: List[str] = []
        if resolver_variant == "legacy":
            parts.append(
                "pip's legacy dependency resolver does not consider dependency "
                "conflicts when selecting packages. This behaviour is the "
                "source of the following dependency conflicts."
            )
        else:
            assert resolver_variant == "resolvelib"
            parts.append(
                "pip's dependency resolver does not currently take into account "
                "all the packages that are installed. This behaviour is the "
                "source of the following dependency conflicts."
            )

        # NOTE: There is some duplication here, with commands/check.py
        for project_name in missing:
            version = package_set[project_name][0]
            for dependency in missing[project_name]:
                message = (
                    f"{project_name} {version} requires {dependency[1]}, "
                    "which is not installed."
                )
                parts.append(message)

        for project_name in conflicting:
            version = package_set[project_name][0]
            for dep_name, dep_version, req in conflicting[project_name]:
                message = (
                    "{name} {version} requires {requirement}, but {you} have "
                    "{dep_name} {dep_version} which is incompatible."
                ).format(
                    name=project_name,
                    version=version,
                    requirement=req,
                    dep_name=dep_name,
                    dep_version=dep_version,
                    you=("you" if resolver_variant == "resolvelib" else "you'll"),
                )
                parts.append(message)

        logger.critical("\n".join(parts))


def get_lib_location_guesses(
    user: bool = False,
    home: Optional[str] = None,
    root: Optional[str] = None,
    isolated: bool = False,
    prefix: Optional[str] = None,
) -> List[str]:
    scheme = get_scheme(
        "",
        user=user,
        home=home,
        root=root,
        isolated=isolated,
        prefix=prefix,
    )
    return [scheme.purelib, scheme.platlib]


def site_packages_writable(root: Optional[str], isolated: bool) -> bool:
    return all(
        test_writable_dir(d)
        for d in set(get_lib_location_guesses(root=root, isolated=isolated))
    )


def decide_user_install(
    use_user_site: Optional[bool],
    prefix_path: Optional[str] = None,
    target_dir: Optional[str] = None,
    root_path: Optional[str] = None,
    isolated_mode: bool = False,
) -> bool:
    """Determine whether to do a user install based on the input options.

    If use_user_site is False, no additional checks are done.
    If use_user_site is True, it is checked for compatibility with other
    options.
    If use_user_site is None, the default behaviour depends on the environment,
    which is provided by the other arguments.
    """
    # In some cases (config from tox), use_user_site can be set to an integer
    # rather than a bool, which 'use_user_site is False' wouldn't catch.
    if (use_user_site is not None) and (not use_user_site):
        logger.debug("Non-user install by explicit request")
        return False

    if use_user_site:
        if prefix_path:
            raise CommandError(
                "Can not combine '--user' and '--prefix' as they imply "
                "different installation locations"
            )
        if virtualenv_no_global():
            raise InstallationError(
                "Can not perform a '--user' install. User site-packages "
                "are not visible in this virtualenv."
            )
        logger.debug("User install by explicit request")
        return True

    # If we are here, user installs have not been explicitly requested/avoided
    assert use_user_site is None

    # user install incompatible with --prefix/--target
    if prefix_path or target_dir:
        logger.debug("Non-user install due to --prefix or --target option")
        return False

    # If user installs are not enabled, choose a non-user install
    if not site.ENABLE_USER_SITE:
        logger.debug("Non-user install because user site-packages disabled")
        return False

    # If we have permission for a non-user install, do that,
    # otherwise do a user install.
    if site_packages_writable(root=root_path, isolated=isolated_mode):
        logger.debug("Non-user install because site-packages writeable")
        return False

    logger.info(
        "Defaulting to user installation because normal site-packages "
        "is not writeable"
    )
    return True


def create_os_error_message(
    error: OSError, show_traceback: bool, using_user_site: bool
) -> str:
    """Format an error message for an OSError

    It may occur anytime during the execution of the install command.
    """
    parts = []

    # Mention the error if we are not going to show a traceback
    parts.append("Could not install packages due to an OSError")
    if not show_traceback:
        parts.append(": ")
        parts.append(str(error))
    else:
        parts.append(".")

    # Spilt the error indication from a helper message (if any)
    parts[-1] += "\n"

    # Suggest useful actions to the user:
    #  (1) using user site-packages or (2) verifying the permissions
    if error.errno == errno.EACCES:
        user_option_part = "Consider using the `--user` option"
        permissions_part = "Check the permissions"

        if not running_under_virtualenv() and not using_user_site:
            parts.extend(
                [
                    user_option_part,
                    " or ",
                    permissions_part.lower(),
                ]
            )
        else:
            parts.append(permissions_part)
        parts.append(".\n")

    # Suggest the user to enable Long Paths if path length is
    # more than 260
    if (
        WINDOWS
        and error.errno == errno.ENOENT
        and error.filename
        and len(error.filename) > 260
    ):
        parts.append(
            "HINT: This error might have occurred since "
            "this system does not have Windows Long Path "
            "support enabled. You can find information on "
            "how to enable this at "
     2.2.1",
		"32.2.2",
		"32.2.3",
		"32.2.4",
		"32.2.5",
		"32.2.6",
		"32.2.7",
		"32.2.8",
		"32.3.0",
		"32.3.1",
		"32.3.2",
		"32.3.3"
	],
	"129.0.6668.0": [
		"33.0.0-alpha.1"
	],
	"130.0.6672.0": [
		"33.0.0-alpha.2",
		"33.0.0-alpha.3",
		"33.0.0-alpha.4",
		"33.0.0-alpha.5",
		"33.0.0-alpha.6",
		"33.0.0-beta.1",
		"33.0.0-beta.2",
		"33.0.0-beta.3",
		"33.0.0-beta.4"
	],
	"130.0.6723.19": [
		"33.0.0-beta.5",
		"33.0.0-beta.6",
		"33.0.0-beta.7"
	],
	"130.0.6723.31": [
		"33.0.0-beta.8",
		"33.0.0-beta.9",
		"33.0.0-beta.10"
	],
	"130.0.6723.44": [
		"33.0.0-beta.11",
		"33.0.0"
	],
	"130.0.6723.59": [
		"33.0.1",
		"33.0.2"
	],
	"130.0.6723.91": [
		"33.1.0"
	],
	"130.0.6723.118": [
		"33.2.0"
	],
	"130.0.6723.137": [
		"33.2.1"
	],
	"130.0.6723.152": [
		"33.3.0"
	],
	"130.0.6723.170": [
		"33.3.1"
	],
	"130.0.6723.191": [
		"33.3.2",
		"33.4.0",
		"33.4.1",
		"33.4.2",
		"33.4.3",
		"33.4.4",
		"33.4.5",
		"33.4.6",
		"33.4.7",
		"33.4.8",
		"33.4.9",
		"33.4.10",
		"33.4.11"
	],
	"131.0.6776.0": [
		"34.0.0-alpha.1"
	],
	"132.0.6779.0": [
		"34.0.0-alpha.2"
	],
	"132.0.6789.1": [
		"34.0.0-alpha.3",
		"34.0.0-alpha.4",
		"34.0.0-alpha.5",
		"34.0.0-alpha.6",
		"34.0.0-alpha.7"
	],
	"132.0.6820.0": [
		"34.0.0-alpha.8"
	],
	"132.0.6824.0": [
		"34.0.0-alpha.9",
		"34.0.0-beta.1",
		"34.0.0-beta.2",
		"34.0.0-beta.3"
	],
	"132.0.6834.6": [
		"34.0.0-beta.4",
		"34.0.0-beta.5"
	],
	"132.0.6834.15": [
		"34.0.0-beta.6",
		"34.0.0-beta.7",
		"34.0.0-beta.8"
	],
	"132.0.6834.32": [
		"34.0.0-beta.9",
		"34.0.0-beta.10",
		"34.0.0-beta.11"
	],
	"132.0.6834.46": [
		"34.0.0-beta.12",
		"34.0.0-beta.13"
	],
	"132.0.6834.57": [
		"34.0.0-beta.14",
		"34.0.0-beta.15",
		"34.0.0-beta.16"
	],
	"132.0.6834.83": [
		"34.0.0",
		"34.0.1"
	],
	"132.0.6834.159": [
		"34.0.2"
	],
	"132.0.6834.194": [
		"34.1.0",
		"34.1.1"
	],
	"132.0.6834.196": [
		"34.2.0"
	],
	"132.0.6834.210": [
		"34.3.0",
		"34.3.1",
		"34.3.2",
		"34.3.3",
		"34.3.4",
		"34.4.0",
		"34.4.1",
		"34.5.0",
		"34.5.1",
		"34.5.2",
		"34.5.3",
		"34.5.4",
		"34.5.5",
		"34.5.6",
		"34.5.7",
		"34.5.8"
	],
	"133.0.6920.0": [
		"35.0.0-alpha.1",
		"35.0.0-alpha.2",
		"35.0.0-alpha.3",
		"35.0.0-alpha.4",
		"35.0.0-alpha.5",
		"35.0.0-beta.1"
	],
	"134.0.6968.0": [
		"35.0.0-beta.2",
		"35.0.0-beta.3",
		"35.0.0-beta.4"
	],
	"134.0.6989.0": [
		"35.0.0-beta.5"
	],
	"134.0.6990.0": [
		"35.0.0-beta.6",
		"35.0.0-beta.7"
	],
	"134.0.6998.10": [
		"35.0.0-beta.8",
		"35.0.0-beta.9"
	],
	"134.0.6998.23": [
		"35.0.0-beta.10",
		"35.0.0-beta.11",
		"35.0.0-beta.12"
	],
	"134.0.6998.44": [
		"35.0.0-beta.13",
		"35.0.0",
		"35.0.1"
	],
	"134.0.6998.88": [
		"35.0.2",
		"35.0.3"
	],
	"134.0.6998.165": [
		"35.1.0",
		"35.1.1"
	],
	"134.0.6998.178": [
		"35.1.2"
	],
	"134.0.6998.179": [
		"35.1.3",
		"35.1.4",
		"35.1.5"
	],
	"134.0.6998.205": [
		"35.2.0",
		"35.2.1",
		"35.2.2",
		"35.3.0",
		"35.4.0",
		"35.5.0",
		"35.5.1",
		"35.6.0",
		"35.7.0",
		"35.7.1",
		"35.7.2",
		"35.7.4",
		"35.7.5"
	],
	"135.0.7049.5": [
		"36.0.0-alpha.1"
	],
	"136.0.7062.0": [
		"36.0.0-alpha.2",
		"36.0.0-alpha.3",
		"36.0.0-alpha.4"
	],
	"136.0.7067.0": [
		"36.0.0-alpha.5",
		"36.0.0-alpha.6",
		"36.0.0-beta.1",
		"36.0.0-beta.2",
		"36.0.0-beta.3",
		"36.0.0-beta.4"
	],
	"136.0.7103.17": [
		"36.0.0-beta.5"
	],
	"136.0.7103.25": [
		"36.0.0-beta.6",
		"36.0.0-beta.7"
	],
	"136.0.7103.33": [
		"36.0.0-beta.8",
		"36.0.0-beta.9"
	],
	"136.0.7103.48": [
		"36.0.0",
		"36.0.1"
	],
	"136.0.7103.49": [
		"36.1.0",
		"36.2.0"
	],
	"136.0.7103.93": [
		"36.2.1"
	],
	"136.0.7103.113": [
		"36.3.0",
		"36.3.1"
	],
	"136.0.7103.115": [
		"36.3.2"
	],
	"136.0.7103.149": [
		"36.4.0"
	],
	"136.0.7103.168": [
		"36.5.0"
	],
	"136.0.7103.177": [
		"36.6.0",
		"36.7.0",
		"36.7.1",
		"36.7.3",
		"36.7.4",
		"36.8.0",
		"36.8.1",
		"36.9.0",
		"36.9.1",
		"36.9.2",
		"36.9.3",
		"36.9.4",
		"36.9.5"
	],
	"137.0.7151.0": [
		"37.0.0-alpha.1",
		"37.0.0-alpha.2"
	],
	"138.0.7156.0": [
		"37.0.0-alpha.3"
	],
	"138.0.7165.0": [
		"37.0.0-alpha.4"
	],
	"138.0.7177.0": [
		"37.0.0-alpha.5"
	],
	"138.0.7178.0": [
		"37.0.0-alpha.6",
		"37.0.0-alpha.7",
		"37.0.0-beta.1",
		"37.0.0-beta.2"
	],
	"138.0.7190.0": [
		"37.0.0-beta.3"
	],
	"138.0.7204.15": [
		"37.0.0-beta.4",
		"37.0.0-beta.5",
		"37.0.0-beta.6",
		"37.0.0-beta.7"
	],
	"138.0.7204.23": [
		"37.0.0-beta.8"
	],
	"138.0.7204.35": [
		"37.0.0-beta.9",
		"37.0.0",
		"37.1.0"
	],
	"138.0.7204.97": [
		"37.2.0",
		"37.2.1"
	],
	"138.0.7204.100": [
		"37.2.2",
		"37.2.3"
	],
	"138.0.7204.157": [
		"37.2.4"
	],
	"138.0.7204.168": [
		"37.2.5"
	],
	"138.0.7204.185": [
		"37.2.6"
	],
	"138.0.7204.224": [
		"37.3.0"
	],
	"138.0.7204.235": [
		"37.3.1"
	],
	"138.0.7204.243": [
		"37.4.0"
	],
	"138.0.7204.251": [
		"37.5.0",
		"37.5.1",
		"37.6.0",
		"37.6.1",
		"37.7.0",
		"37.7.1",
		"37.8.0",
		"37.9.0",
		"37.10.0",
		"37.10.1",
		"37.10.2",
		"37.10.3"
	],
	"139.0.7219.0": [
		"38.0.0-alpha.1",
		"38.0.0-alpha.2",
		"38.0.0-alpha.3"
	],
	"140.0.7261.0": [
		"38.0.0-alpha.4",
		"38.0.0-alpha.5",
		"38.0.0-alpha.6"
	],
	"140.0.7281.0": [
		"38.0.0-alpha.7",
		"38.0.0-alpha.8"
	],
	"140.0.7301.0": [
		"38.0.0-alpha.9"
	],
	"140.0.7309.0": [
		"38.0.0-alpha.10"
	],
	"140.0.7312.0": [
		"38.0.0-alpha.11"
	],
	"140.0.7314.0": [
		"38.0.0-alpha.12",
		"38.0.0-alpha.13",
		"38.0.0-beta.1"
	],
	"140.0.7327.0": [
		"38.0.0-beta.2",
		"38.0.0-beta.3"
	],
	"140.0.7339.2": [
		"38.0.0-beta.4",
		"38.0.0-beta.5",
		"38.0.0-beta.6"
	],
	"140.0.7339.16": [
		"38.0.0-beta.7"
	],
	"140.0.7339.24": [
		"38.0.0-beta.8",
		"38.0.0-beta.9"
	],
	"140.0.7339.41": [
		"38.0.0-beta.11",
		"38.0.0"
	],
	"140.0.7339.80": [
		"38.1.0"
	],
	"140.0.7339.133": [
		"38.1.1",
		"38.1.2",
		"38.2.0",
		"38.2.1",
		"38.2.2"
	],
	"140.0.7339.240": [
		"38.3.0",
		"38.4.0"
	],
	"140.0.7339.249": [
		"38.5.0",
		"38.6.0",
		"38.7.0",
		"38.7.1",
		"38.7.2",
		"38.8.0",
		"38.8.1",
		"38.8.2",
		"38.8.4"
	],
	"141.0.7361.0": [
		"39.0.0-alpha.1",
		"39.0.0-alpha.2"
	],
	"141.0.7390.7": [
		"39.0.0-alpha.3",
		"39.0.0-alpha.4",
		"39.0.0-alpha.5"
	],
	"142.0.7417.0": [
		"39.0.0-alpha.6",
		"39.0.0-alpha.7",
		"39.0.0-alpha.8",
		"39.0.0-alpha.9",
		"39.0.0-beta.1",
		"39.0.0-beta.2",
		"39.0.0-beta.3"
	],
	"142.0.7444.34": [
		"39.0.0-beta.4",
		"39.0.0-beta.5"
	],
	"142.0.7444.52": [
		"39.0.0"
	],
	"142.0.7444.59": [
		"39.1.0",
		"39.1.1"
	],
	"142.0.7444.134": [
		"39.1.2"
	],
	"142.0.7444.162": [
		"39.2.0",
		"39.2.1",
		"39.2.2"
	],
	"142.0.7444.175": [
		"39.2.3"
	],
	"142.0.7444.177": [
		"39.2.4",
		"39.2.5"
	],
	"142.0.7444.226": [
		"39.2.6"
	],
	"142.0.7444.235": [
		"39.2.7"
	],
	"142.0.7444.265": [
		"39.3.0",
		"39.4.0",
		"39.5.0",
		"39.5.1",
		"39.5.2",
		"39.6.0",
		"39.6.1",
		"39.7.0"
	],
	"143.0.7499.0": [
		"40.0.0-alpha.2"
	],
	"144.0.7506.0": [
		"40.0.0-alpha.4"
	],
	"144.0.7526.0": [
		"40.0.0-alpha.5",
		"40.0.0-alpha.6",
		"40.0.0-alpha.7",
		"40.0.0-alpha.8"
	],
	"144.0.7527.0": [
		"40.0.0-beta.1",
		"40.0.0-beta.2"
	],
	"144.0.7547.0": [
		"40.0.0-beta.3",
		"40.0.0-beta.4",
		"40.0.0-beta.5"
	],
	"144.0.7559.31": [
		"40.0.0-beta.6",
		"40.0.0-beta.7",
		"40.0.0-beta.8"
	],
	"144.0.7559.60": [
		"40.0.0-beta.9",
		"40.0.0"
	],
	"144.0.7559.96": [
		"40.1.0"
	],
	"144.0.7559.111": [
		"40.2.0",
		"40.2.1"
	],
	"144.0.7559.134": [
		"40.3.0",
		"40.4.0"
	],
	"144.0.7559.173": [
		"40.4.1"
	],
	"144.0.7559.177": [
		"40.5.0",
		"40.6.0"
	],
	"144.0.7559.220": [
		"40.6.1"
	],
	"146.0.7635.0": [
		"41.0.0-alpha.1",
		"41.0.0-alpha.2"
	],
	"146.0.7645.0": [
		"41.0.0-alpha.3"
	],
	"146.0.7650.0": [
		"41.0.0-alpha.4",
		"41.0.0-alpha.5",
		"41.0.0-alpha.6",
		"41.0.0-beta.1",
		"41.0.0-beta.2",
		"41.0.0-beta.3"
	],
	"146.0.7666.0": [
		"41.0.0-beta.4"
	],
	"146.0.7680.16": [
		"41.0.0-beta.5",
		"41.0.0-beta.6"
	],
	"146.0.7680.31": [
		"41.0.0-beta.7"
	]
};
